### ISSUE TITLE
useLocalStorage sometimes returns default value when it should not

### ISSUE THREAD

**Issue by sudall** [mantinedev/mantine#5756]
### Dependencies check up

- [X] I have verified that I use latest version of all @mantine/* packages

### What version of @mantine/* packages do you have in package.json?

7.5.2

### What package has an issue?

@mantine/hooks

### What framework do you use?

Next.js

### In which browsers you can reproduce the issue?

Chrome

### Describe the bug

useLocalStorage sometimes returns default value when there's a valid value in local storage. I found this problematic for the situation of setting initial values for a form from local storage. ie. there's a value in local storage, but the hook is giving me the default value on the first one to three renders and eventually gives me the correct value from storage.

I'm not sure how to repro this other than it seems to happen when I open a modal dialog with the useLocalStorage hook in it, as that's where my form is.

Here's the output of some simple logging to prove it.
![image](https://github.com/mantinedev/mantine/assets/12477686/37ec3332-d21d-4261-a7b8-923249a7bea3)
![image](https://github.com/mantinedev/mantine/assets/12477686/1a6fca80-b56d-48f1-9fb4-0d65f8d4a617)

### Self-service

- [ ] I would be willing to implement a fix for this issue

**Comment by rtivital** [comment]
If you want to initialize state variable with local storage value, set `getInitialValueInEffect: false` – https://mantine.dev/hooks/use-local-storage/#definition

**Comment by sudall** [comment]
Ah thank you. I must have missed that option. Just tried and it's working. What's the use case for it being `true`?

**Comment by Pirulax** [comment]
Id would be nice to improve the docs on `getInitialValueInEffect`.
I spent quite a bit of time trying to figure out why my layout with mantine react table kept being reset... Turns out `useLocalStorage` was loading the values with a delay (the same issue you had).
Had the docs been clearer on this matter (eg.: that disabling this feature will make this hook immediately return the stored value) would've been very helpful.

**Comment by am-burban** [comment]
For future reference (I found this issue through Google), it would be really good to know the answer to this question:

> What's the use case for it being true?

I'd be happy to open a PR to improve the docs, but I think that information is crucial and I have no idea :-/

**Comment by sudall** [comment]
> For future reference (I found this issue through Google), it would be really good to know the answer to this question:
> 
> > What's the use case for it being true?
> 
> I'd be happy to open a PR to improve the docs, but I think that information is crucial and I have no idea 🫤

I think I've finally stumbled upon why. I think it's a convenience for server-side rendering. When you try to use local storage in a framework that uses SSR (like NextJS), the server will render not knowing about what's in local storage on the client. This ends up causing a render mismatch between client and server and NextJS gives you an error in dev. At the moment, I think I still prefer `getInitialValueInEffect` to be off because of the bug I described and because having it off shows me where I should only do client rendering. Perhaps there is a way to have it so that when a value is requested from localstorage on the client, it always uses the actual value, and then uses the default value on the server.
